tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,344 @@
+          tmp_output = at::redispatch::{functional_api_name}(
+            {', '.join([keyset] + [arg_name(a) for a in f.func.arguments.flat_non_out])});
+      }}
+      {mutable_input_post_processing}
+      {maybe_return}{return_names_str(f)};"""
+
+
+
+def emit_registration(f: NativeFunction, native_api_name: Optional[str]) -> str:
+    if f.has_composite_implicit_autograd_kernel:
+        assert native_api_name is not None, f"Found a view op ({f.func.name}) with no cpu or composite kernel." \
+            f" The functionalization pass codegen needs to know the at::native::{f.func.name} kernel name of every view op," \
+            f" which is currently does by looking up the cpu/composite kernel kernel name of the op. If you encountered a view op" \
+            f" without a cpu/composite kernel, you'll need to change this logic"
+        sig = DispatcherSignature.from_schema(f.func)
+        # Note [Composite view ops in the functionalization pass]
+        # We don't need to worry about implemententing functionalization kernels for views with
+        # CompositeImplicitAutograd kernels, because we can just decompose them into their base operators.
+        # We can't just opt the entire Functionalization dispatch key into the composite keyset though,
+        # because we don't want to decompose non-view ops that are composite, like `at::ones`.
+        registration_str = f'static_cast<{sig.ptr_type()}>(at::native::{native_api_name})'
+    else :
+        registration_str = f'TORCH_FN(functionalization::{type_wrapper_name(f)})'
+
+    return f"""
+m.impl("{f.func.name}",
+       {registration_str}
+);
+"""
+
+def emit_definition(f: NativeFunction, g: Optional[NativeFunctionsGroup], native_api_name: Optional[str]) -> str:
+    # order is important here, ops that are both views and mutations should hit the view path.
+    if is_view_op(f):
+        assert native_api_name is not None, f"Found a view op ({f.func.name}) with no cpu or composite kernel." \
+            f" The functionalization pass codegen needs to know the at::native::{f.func.name} kernel name of every view op," \
+            f" which is currently does by looking up the cpu/composite kernel kernel name of the op. If you encountered a view op" \
+            f" without a cpu/composite kernel, you'll need to change this logic"
+        body_str = emit_view_functionalization_body(f, g, native_api_name)
+    else:
+        body_str = emit_inplace_functionalization_body(f, g)
+    return f"""
+{cpp.returns_type(f.func.returns).cpp_type()} {type_wrapper_name(f)}({gen_formals(f)}) {{
+  {body_str}
+}}
+"""
+
+def emit_declaration_for_noncomposite_views(f: NativeFunction) -> str:
+    # For every view op, we need a corresponding "inverse view" function.
+    # This generates the declarations so we get a good compiler error when someone adds a new view.
+    for r in f.func.returns:
+        assert r.type.is_tensor_like()
+
+    name = dispatcher.name(f.func)
```

Comment:
You've set up a calling convention here: inverse functions are defined by the user, predeclared here, and called in the functionalization body above. So you should actually explicitly say so, rather than have the calling convention be implicitly encoded by the manipulations here and above. It's not such a big deal since the calling convention is limited to this file, but it is generally good style to make a fresh api module for this. Then it would have a few functions like `name` and `arguments`, and you can also make a `ViewInverseSignature` class for it.

Reply:
This would also make `skip_self` better motivated: there's never a situation in *dispatcher API* that you want to skip the self argument; this entirely exists for view inverses (since you need to treat the self argument specially). So you would probably just reimplement `inverse.arguments` from scratch. It also would make it clear that the signature here matching the dispatcher API is a faux amis; really it's its own thing